doc/specs/#476 - Package Pinning.md

```diff
@@ -0,0 +1,205 @@
+- User may want some of the packages to stay in some major versions but allow minor version changes during upgrade.
+
+## Solution Design
+
+#### Package Pinning types
+To achieve goals listed above, winget will support 3 types of Package Pinning:
+- **Blocking:** The package is blocked from `winget upgrade --all` or `winget upgrade <specific package>`, user has to unblock the package to let winget perform upgrade.
+- **Pinning:** The package is excluded from `winget upgrade --all` but allowed in `winget upgrade <specific package>`, a new argument `--include-pinned` will be introduced to let `winget upgrade --all` to include pinned packages.
+- **Gating:** The package is pinned to specific version(s). For example, if a package is pinned to version `1.2.*`, any version between `1.2.0` to `1.2.<anything>` is considered valid.
+
+To allow user override, `--force` can be used with `winget upgrade <specific package>` to override some of the pinning created above.
+
+#### Package Pinning Configuration Storage
+
+A separate sqlite db (other than the existing tracking catalog) will be created to store the package pinning configurations from user.
+```text
+PackageIdentifier       SourceIdentifier       Version        PinningType
+----------------------------------------------------------------------------
+Microsoft.TestApp       winget                 1.2.*          Gating
+```
+
+**Notes:** For this iteration, winget will only support pinning packages that are locally installed and correlatable with at least one of the remote sources. Winget will record a pinned package by the PackageIdentifier and SourceIdentifier. There can only be one pinning configuration for a specific package. In the future, winget may consider pinning packages from installed packages (upon improving the installed package's PackageIdentifier logic).
+
+## UI/UX Design
+
+#### winget pin commands
+
+A new `winget pin` command with 3 sub-commands will be introduced.
+- Add package pinning configuration:
+  
+  `winget pin add <package> [--version <optional gated version>] [--source <source>] [--force] [--blocking]`
+  
+- Remove package pinning configuration:
+
+  `winget pin remove <package> [--source <source>] [--force]`
+  
+- List package pinning configuration: 
+  
+  `winget pin list <package> [--source <source>]` for a specific package or `winget pin list` to list all
+
+#### Blocking
+To block a package from upgrade, use `winget pin add <package> --blocking`
+```text
+cmd> winget pin Microsoft.TestApp --blocking
+```
+Now the pinning configuration is recorded as
+```text
+PackageIdentifier       SourceIdentifier       Version        PinningType
+----------------------------------------------------------------------------
+Microsoft.TestApp       winget                                Blocking
+Microsoft.TestAppStore  msstore                               Blocking
```

Comment:
If the pins are recorded for each source present at the time of adding the pin then if a user later adds another source the package would be upgraded from there right? I think from a UX perspective if I ran `winget pin add <package> --blocking` I would expect the package to be pinned indefinitely for all current and future sources. So I would prefer if pinning a package without explicitly listing individual sources would result in a different, bool SQLite column being set like `AllSources` rather than multiple entries with arbitrary `SourceIdentifier`s.

Reply:
Re @Trenly , all these pin configurations only apply to winget behaviors, we cannot control upgrades outside winget. Blocking simply blocks winget from doing anything to the blocked package. In the original issue, some community member wanted winget to block upgrade for some specific package, because they want to perform upgrades outside winget through other channel.

Re @jantari , yes, it would be ideal to pin an installed package for all future sources. Due to current limitation on the package identifier generation for installed packages, pinning from installed source is put into future considerations. Currently, a package identifier for an installed package is its produce code, if user upgrade the package, the pin is no longer in effect since most likely product code will change after an upgrade. Regarding `AllSources` in sqlite, it's trivial to record it, but we don't have efficient cross reference for remote sources yet, so it's unfortunately moved as future considerations.